### ISSUE TITLE
crypto/x509: Update test case in verify_test

### ISSUE THREAD

**Issue by SparrowLii** [golang/go#40604]
### What version of Go are you using (`go version`)?

go version go1.14.6 

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

windows/amd64

### What did you do?

I just run the test instruction in my desktop (go test crypto/x509) .

### What did you expect to see?

get the right certificate chain in the test.

### What did you see instead?

There was an error followed: 

--- FAIL: TestSystemVerify (0.03s)
--- FAIL: TestSystemVerify/SHA-384 (0.00s)
verify_test.go:584: no expected chain matched BR/Moip Pagamentos S.A./MOIP,SSL Blindado EV/api.moip.com.br -> GB/COMODO CA Limited//COMODO RSA Extended Validation Secure Server CA -> GB/COMODO CA Limited//COMODO RSA Certification Authority -> GB/Comodo CA Limited//AAA Certificate Services

The expected chain in the verify_test is:
expectedChains: [][]string{
{
"api.moip.com.br",
"COMODO RSA Extended Validation Secure Server CA",
"COMODO RSA Certification Authority",
"AddTrust External CA Root",
},
},

I'm wondering if this is a wrong of mine alone or we should update the test case either.

**Comment by KCoen** [comment]
The root certificate used for this test "AddTrust External CA Root" Expired on 2020-05-30, the test tries to override the time, but this is ignored for root certificates on windows.

I presume this test would succeed on windows if the root certificate was valid.

**Comment by hyangah** [comment]
cc @golang/osp-team how can we check whether the builders run these tests and if so, why they didn't fail?

**Comment by SparrowLii** [comment]
I tested crypto/x509 in several different desktops and found that only those PCs which installed Windows after 2020-5-30 would fail. By the way, I has update my go version to 1.15.
I guess it dose be because "AddTrust External CA Root" Expired, as @KCoen said.

**Comment by dmitshur** [comment]
> how can we check whether the builders run these tests

@hyangah It will be possible to click on the "ok" text at build.golang.org to see successful build logs after #34119 and #34744 are resolved. It's already implemented in my local prototype, which I need to finish. Please subscribe to those issues for updates.

Until then, it needs to be accessed manually. See [here](https://build.golang.org/log/8fa94478b5bd4aeecbb0288915a6aa0ce860b3bc) for the build log from a recent windows-amd64-2016 builder result. It contains:

```
windows-amd64-2016 at c2e73fb446bffd02c651e51c6641cc90fd065b70

:: Running C:\workdir\go\src\make.bat with args [...]

[...]
ok  	crypto/x509	1.434s
[...]
```

> if so, why they didn't fail?

The most likely explanation is there is some difference in environment between the builder (its environment variables, Windows image, etc.).

Our latest Windows builder images are from 2016 (Windows Server 2016), so based on @SparrowLii's comment, it's possible this would be caught on a builder with a newer Windows image. /cc @andybons @toothrot @cagedmantis @valsorda

**Comment by alexbrainman** [comment]
This happens on my computer too. My computer is standard Windows 10. It gets updated automatically every few month without me even noticing. So this would be the same for most Windows 10 users.

I also googled for `no expected chain matched BR/Moip Pagamentos`,  and I found related issues #7824 and #11730. I did not investigate, if they are related or not.

I don't run all.bat regularly. But I am making runtime change, so I run all.bat and noticed this issue.

Let me know, if you need me do bisect this issue.

Thank you.

Alex

**Comment by rolandshoemaker** [comment]
Certainly possible that if the builders are using a static Windows image that they aren't pulling changes to the authroot list which may cause the root store to be in a different state than on actual machines?

The [`CertGetCertificateChain ` docs](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certgetcertificatechain) are slightly vague about how `pTime` (which is what we use `VerifyOptions.CurrentTime` for) is used for root validation:
> Note that the time does not affect trust list, revocation, or root store checking. ... Trust in a particular certificate being a trusted root is based on the current state of the root store and not the state of the root store at a time passed in by this parameter.

It could be that `CertGetCertificateChain` won't pass back a expired root even if it is still in the authroot list and `pTime` is set to a time when it would be valid...

**Comment by KCoen** [comment]
So on machines with both "AddTrust External CA Root" and "AAA Certificate Services" root certs installed, windows will not return the AAA cert over the AddTrust cert.

I found that calling CertGetCertificateChain with CERT_CHAIN_RETURN_LOWER_QUALITY_CONTEXTS on machines with both installed, will force windows to keep looking and it will find it eventually and pass it back in a different field.

I've submitted changes(?) for that here https://go-review.googlesource.com/c/go/+/257257 

Eventually machines might not have the original root that the test is trying to find and it'll break again, it might just be a better idea to split out the tests that are ran against a system certificate store, because testing the entire chain against a preset is just gonna keep breaking every couple months.

**Comment by FiloSottile** [comment]
@rolandshoemaker you seemed to have looked into this, so I'll leave it to you, but let me know if you have any issue with it.

**Comment by egroj97** [comment]
Hello, tried to install from source, but when testing, I get the same error:

```
verify test.go:584: no expected chain matched BR/Moip Pagamentos S.A./MOIP,SSL Blindado EV/api.moip.com.br -> GB/COMODO CA Limited//COMODO RSA Extended Validation Secure Server CA -> GB/COMODO CA Limited//COMODO RSA Certification Authority
```

I'm building from branch: `go1.15.6`, the error is still showing because it hasn't been added to an official release?

**Comment by FiloSottile** [comment]
Correct, this is fixed in Go 1.16 but not backported to Go 1.15. If you are just trying to build Go, you can ignore the failure and use make.bash. I'm not sure what our policy on fixing tests in old branches is. It would be ok to just skip this test. (The actual fix in Go 1.16 is too large to backport.) /cc @golang/release
